syncer/syncer.go

```diff
@@ -366,33 +367,37 @@ func (s *Syncer) getLayerFromPeers(ctx context.Context, layerID types.LayerID) (
 	ch := s.fetcher.PollLayerHash(ctx, layerID)
 	hashRes := <-ch
 	if hashRes.Err != nil {
-		return nil, fmt.Errorf("PollLayerHash: %w", hashRes.Err)
+		return nil, fmt.Errorf("getLayerFromPeers: %w", hashRes.Err)
 	}
 	// TODO: resolve hash with peers
 	hashes := make(map[types.Hash32][]peers.Peer)
 	for lyrHash, peers := range hashRes.Hashes {
 		hashes[lyrHash.Hash] = peers
 	}
 	bch := s.fetcher.PollLayerBlocks(ctx, layerID, hashes)
-	res := <-bch
-	if res.Err != nil {
-		if res.Err == layerfetcher.ErrZeroLayer {
-			return types.NewLayer(layerID), nil
+	select {
+	case res := <-bch:
+		if res.Err != nil {
+			if res.Err == layerfetcher.ErrZeroLayer {
+				return types.NewLayer(layerID), nil
+			}
+			return nil, fmt.Errorf("getLayerFromPeers: %w", res.Err)
 		}
-		return nil, fmt.Errorf("PollLayerBlocks: %w", res.Err)
+	case <-time.After(s.fetcher.GetTimeout()):
```

Comment:
i don't understand why we need this. fetcher/msgserver already has timeout mechanism. does it make sense to adjust the value there instead?

Reply:
I don't see a timeout mechanism. I traced this down several levels:

LayerFetcher.PollLayerBlocks: blocks on MsgServer.SendRequest
MsgServer.SendRequest: blocks on Switch.sendWrappedMessage
Switch.sendWrappedMessage: blocks on Switch.sendMessageImpl
Switch.sendMessageImpl: blocks on conn.Send
FormattedConnection.Send: blocks on writing data to a channel

What am I missing?

Reply:
https://github.com/spacemeshos/go-spacemesh/blob/07f9602c2f06094180a84962722116ba486c9cd6/p2p/server/msgserver.go#L216

Reply:
@lrettig

Reply:
That's a callback, but it doesn't change the stack I described above: the request itself still blocks. It's possible that the error callback gets called while the request goroutine is still blocking. This should be cleaned up - to be clear, what's being done in this PR is just a messy workaround.

Reply:
In other words, `MessageServer.cleanStaleMessages` does not unblock the request goroutine

Reply:
after RequestTimeout elapsed, the end result is that  an error will be delivered to `bch` channel and unblock.
the code in layerfetcher uses the same handler for good data and error, and the responses got aggregated and sent to `bch` channel. the new timeout code just added another level of timeout.

Reply:
I see where the error callback gets called. But I don't see where the original request gets unblocked. Could you explain this part to me?

Reply:
https://github.com/spacemeshos/go-spacemesh/blob/389ab13e4270364f1a892f90f003ca20da92ab9e/layerfetcher/layers.go#L416
is the error callback for this block IDs request `SendRequest()`

https://github.com/spacemeshos/go-spacemesh/blob/389ab13e4270364f1a892f90f003ca20da92ab9e/layerfetcher/layers.go#L485
is the gorountine launched after all expected peers responded and all blocks fetched.

https://github.com/spacemeshos/go-spacemesh/blob/389ab13e4270364f1a892f90f003ca20da92ab9e/layerfetcher/layers.go#L521
is where all the `bch`s for the same layer hash are unblocked.

Reply:
`:s/for the same layer hash/for the same layer ID/` in my last sentence

Reply:
I don't think we're on the same page here. These are two different levels of abstraction. You're describing what happens once the callbacks are called inside the message server. I'm pointing out that the message server itself may block on a stuck connection, and may never call those callbacks. Maybe it would be easier to look over the code together on a call?

Reply:
if the msg server itself block on stuck connection, why is the timeout mechanism placed here and not in msg server?

does that mean all clients of msg server need to do its timeout handling? the design is weird that the caller already provide a timeout hander when calling SendRequest yet have to handle the timeout anyway?

can you point me to the line of code where the msg server connection may get stuck?

Reply:
As I mentioned in https://github.com/spacemeshos/go-spacemesh/pull/2463/files#r694103854, it's inside MsgConnection/FormattedConnection. The `Send` method of both of these queues up outbound messages into a channel:

https://github.com/spacemeshos/go-spacemesh/blob/e213df4eb0b87d7c41cd38ba9e622c812d7050c6/p2p/net/conn.go#L281

This will block if the channel is full. The channel is consumed inside `sendListener()`, which passes the outbound messages to `SendSock()`, which blocks until the message is successfully written to the wire:

https://github.com/spacemeshos/go-spacemesh/blob/e213df4eb0b87d7c41cd38ba9e622c812d7050c6/p2p/net/conn.go#L299

This has a timeout, but it's set pretty high, 60 sec by default, and that's per message, so if a connection is totally stuck, then everything will get backed up and block. This is what we saw happen on the old testnet.

#2450 is also part of addressing this, by dropping stuck peers to prevent these things from blocking.

I see your point about weird design, let me take another look and see if there isn't another option.

Reply:
I took a closer look and I think that, with #2450, this timeout is no longer necessary since this call can no longer block. I've removed it. Thanks!